test/src/main/java/org/apache/accumulo/test/functional/TabletServerResourceGroupBalanceIT.java

```diff
@@ -0,0 +1,291 @@
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.accumulo.test.functional;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertNull;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.NoSuchElementException;
+import java.util.Optional;
+import java.util.SortedSet;
+import java.util.TreeSet;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+import org.apache.accumulo.core.Constants;
+import org.apache.accumulo.core.client.Accumulo;
+import org.apache.accumulo.core.client.AccumuloClient;
+import org.apache.accumulo.core.client.admin.NewTableConfiguration;
+import org.apache.accumulo.core.client.admin.TabletHostingGoal;
+import org.apache.accumulo.core.clientImpl.ClientContext;
+import org.apache.accumulo.core.clientImpl.ClientTabletCache;
+import org.apache.accumulo.core.clientImpl.ClientTabletCache.LocationNeed;
+import org.apache.accumulo.core.conf.Property;
+import org.apache.accumulo.core.data.Range;
+import org.apache.accumulo.core.data.TableId;
+import org.apache.accumulo.core.fate.zookeeper.ZooCache;
+import org.apache.accumulo.core.fate.zookeeper.ZooCache.ZcStat;
+import org.apache.accumulo.core.lock.ServiceLock;
+import org.apache.accumulo.core.lock.ServiceLockData;
+import org.apache.accumulo.core.metadata.schema.Ample;
+import org.apache.accumulo.core.metadata.schema.TabletMetadata;
+import org.apache.accumulo.core.metadata.schema.TabletMetadata.Location;
+import org.apache.accumulo.harness.MiniClusterConfigurationCallback;
+import org.apache.accumulo.harness.SharedMiniClusterBase;
+import org.apache.accumulo.minicluster.ServerType;
+import org.apache.accumulo.miniclusterImpl.MiniAccumuloConfigImpl;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.io.Text;
+import org.junit.jupiter.api.AfterAll;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.Test;
+
+import com.google.common.net.HostAndPort;
+
+public class TabletServerResourceGroupBalanceIT extends SharedMiniClusterBase {
```

Comment:
Some more test that could be useful.

 * Test balancing the root and metadata table to diff tserver groups.
 * Test two user tables that balance to diff groups to make sure that works as expected.
 * Test tablets within in a single table that balance to different groups.

Reply:
> Test two user tables that balance to diff groups to make sure that works as expected.

`testBalancerWithResourceGroups` does this. I creates one table in the default group and one table in the GROUP1 group. Then it validates that all of the first tables tablets are located in the tserver in the default group and that all of the second tables tablets are located in the tserver in the GROUP1 group.

Reply:
> Test balancing the root and metadata table to diff tserver groups.

Is this any different than the `testBalancerWithResourceGroups` test? Are these tables handled differently?

Reply:
The root tablet is handled differently, metadata table is more likely to be like everything else. When I added an IT for compaction of root and metadata, I found the root tablet did not work.  Had to add this :

https://github.com/apache/accumulo/blob/406552e46df4412cf8f4f52962446773a9eabee5/server/base/src/main/java/org/apache/accumulo/server/manager/state/ZooTabletStateStore.java#L79-L84

Reply:
We may be able to have the root tablet use the TabletMgmtIterator, but not sure.   Going to open an issue.  It would be nice if it could, then it would be more likely to have consistent behavior.

Reply:
Created #3587

Reply:
@keith-turner - I added balance tests for root and metadata table in 6ca1a88, and they work.

Reply:
> If the SPI supports the functionality, then it would be good to ensure it actually works. 

So, the `TabletBalancer.getResourceGroup` api method works at the tablet level now. The TabletMetadata.getTabletState method returns `TabletState.ASSIGNED_TO_WRONG_GROUP` when a tablet is not hosted on a tablet server that is in the same group that `TabletBalancer.getResourceGroup` returns. `TableLoadBalancer` passes the set of TabletServers that correspond to the resource group to the tables load balancer for assignment and balance purposes. This allows, for example, the SimpleLoadBalancer to assign and balance within the subset of tablet servers.

I think to do what you are asking (assign and balance at the tablet level based on resource group assignment), the table level balancers would need to be modified to make decisions on a per-tablet basis.

Reply:
I guess one way to resolve this is to remove `TabletBalancer.getResourceGroup` such that *only* `TableLoadBalancer` has this functionality. `TabletMetadata.getTabletState` would only check the resource group if the balancer is `TableLoadBalancer`.

Reply:
Discussed the above with @keith-turner , I'm going to merge as-is and create a follow-on issue.